substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java

```diff
@@ -1317,8 +1363,8 @@ private int writeMethodLocation(DebugContext context, PrimaryEntry primaryEntry,
         log(context, "  [0x%08x]     specification  0x%x (%s)", pos, methodSpecOffset, methodKey);
         pos = writeAttrRefAddr(methodSpecOffset, buffer, pos);
         HashMap<DebugLocalInfo, List<Range>> varRangeMap = primary.getVarRangeMap();
-        pos = writeMethodParameterLocations(context, varRangeMap, primary, 2, buffer, pos);
-        pos = writeMethodLocalLocations(context, varRangeMap, primary, 2, buffer, pos);
+        pos = writeMethodParameterLocations(context, null, varRangeMap, primary, 2, buffer, pos);
+        pos = writeMethodLocalLocations(context, null, varRangeMap, primary, 2, buffer, pos);
```

Comment:
The class entry is known in `writeMethodLocation` caller `writeMethodLocations`. Why don't we pass it through?

Reply:
If you look carefully you will notice that this `writeMethodParameterLocations` and `writeMethodLocalLocations` have two callers: `writeMethodDeclaration` and `writeAbstractInlineMethod`. The first caller passes `null` for the class entry because we are writing the params/locals for the original definition of the method. The second caller passes the class entry for the owner of the abstract inline method because we are writing a repeat copy of the locals/params.

In the first case the class of the (original) method is the class whose CU is currently being written. In the second case the class of the (abstract inline) method may not be the same as the class of the CU currently being written.

Passing null in the first case vs the abstract inline class entry is important. The classEntry argument is passed on to `writeMethodParameter/LocalDeclaration` which uses it to parameterize the call to `setMethodLocalIndex`. This setter method indexes values associated with a null class entry differently to those indexed with a non-null class entry because they represent different types of declaration. A value indexed with a null entry identifies a param/local declared under an original declaration. A value indexed with a non-null entry identifies a param/local declared under an abstract inline declaration

If you look at the points where calls to `getMethodLocalIndex` occur you will see that they use the same argument scheme. When defining the location of a param/local for a top level method `null` is passed. When defining the location of a param/local for a (concrete) inlined method the inlined method owner's class entry is passed.

The former step is unchanged from before this patch. The latter step is what ensures that an inline method's param/local location DIEs refer to a param/local specification that exists in the same CU.